src/dosfsck.8
.TH DOSFSCK 8 "2022-08-10" "fatprogs 2.13.0"
.SH NAME
dosfsck \- check and repair MS-DOS file systems
.SH SYNOPSIS
.ad l
.B dosfsck|fsck.msdos|fsck.vfat
.RB [ \-aACflnrtvVwy ]
.RB [ \-d\ \fIpath\fB\ \-d\ \fI...\fB ]
.RB [ \-u\ \fIpath\fB\ \-u\ \fI...\fB ]
.I device
.ad b
.SH DESCRIPTION
.B dosfsck
verifies the consistency of MS-DOS file systems and optionally tries to repair
them. The following file system problems can be corrected (in this order):
.IP "Boot Sector
.PD 0
.IP - 4
Boot sector has invalid parameters. Verify backup boot sector(only for FAT32).
If backup is valid, then replace with it. Otherwise exit.
.IP -
Boot sector does not have filesystem information sector. Create new filesystem
information sector in first sector.
.PD
.IP "FAT"
.PD 0
.IP -
One of FAT has corrupted. they are copied with verified FAT.
.IP -
FAT contains invalid cluster numbers. Cluster is changed to EOF.
.PD
.IP "File/Directory"
.PD 0
.IP -
First entry(.) and second entry(..) are non-directories.
They are set to directory.
.IP -
Dot(.) and dotdot(..) entry's start cluster is invalid.
They are set to valid start cluster.
.IP -
First/second entry is not ".", ".." or empty, create them.
.IP -
File's cluster chain contains a loop. The loop is broken.
.IP -
Bad clusters (read errors). The clusters are marked bad and they are removed
from files owning them. This check is optional.
.IP -
Directories with a large number of bad entries (probably corrupt).
If the number of bad entries is more than 10 times of good entries.
The directory is deleted.
.IP -
Directories "." and ".." in root directory. They are deleted.
.IP -
Bad file names. They can be renamed.
.IP -
Duplicate directory entries. They can be renamed or deleted.
.IP -
Can't determine whether entry is file or directory. Consider it to file.
.IP -
Directories with non-zero size field. Size is set to zero.
.IP -
Directory "." does not point to parent directory. The start pointer is adjusted.
.IP -
Directory ".." does not point to parent of parent directory.
The start pointer is adjusted.
.IP -
Start cluster number of a directory is invalid. The directory is deleted.
.IP -
Volume label has start cluster. Fix it.
.IP -
Volume label is not in root directory. They can be deleted.
.IP -
File contains bad or free clusters. The file is truncated.
.IP -
Two or more files share the same cluster(s).
All but one of the files are truncated. If the file being truncated is
a directory entry that has already been read, the file system check is
restarted after truncation.
.IP -
File's cluster chain is longer than indicated by the size fields.
The file size is changed. (truncation is better?)
.IP -
File's cluster chain is shorter than indicated by the size fields.
The file is truncated.
.PD
.IP "Volume Label"
.PD 0
.IP -
Volume label is not exist. Set to default volume label "NO NAME    ".
.IP -
Volume is invalid. remove it. (set to default label)
.IP -
Multiple volume labels are in root directory entry.
All but one(first found volume entry) of them are removed.
.IP -
One of volume label in boot and root is invalid. Copy valid label to other side.
.PD
.IP "Orphan Cluster"
.PD 0
.IP -
Clusters are marked as used but are not owned by a file.
They are marked as free or created as file (optional).
Files are created in FOUND.NNN directory. where NNN is 3 digit number(000 ~ 999).
Each directory can create 10 thousand entries.
.PD
.LP
Additionally, the following problems are detected, but not repaired or handled:
.PD 0
.IP - 4
in case of the number of FAT is more than 3.
.PD
.LP
When \fBdosfsck\fP checks a file system, it accumulates all changes in memory
and performs them only after all checks are complete.
This can be disabled with the \fB\-w\fP option.
.SH OPTIONS
.IP \fB\-a\fP
Automatically repair the file system. No user intervention is necessary.
Whenever there is more than one method to solve a problem,
the least destructive approach is used.
.IP \fB\-A\fP
Use Atari variation of the MS-DOS filesystem. This is default if
\fBdosfsck\fP is run on an Atari, then this option turns off Atari
format. There are some minor differences in Atari format: Some boot
sector fields are interpreted slightly different, and the special FAT
entries for end-of-file and bad cluster can be different. Under
MS-DOS 0xfff8 is used for EOF and Atari employs 0xffff by default, but
both systems recognize all values from 0xfff8...0xffff as end-of-file.
MS-DOS uses only 0xfff7 for bad clusters, where on Atari values
0xfff0...0xfff7 are for this purpose (but the standard value is still
0xfff7).
.IP \fB\-C\fP
Check only volume dirty flag. If it is set, other options are ignored.
If volume is clean, return 0, otherwise return 4 (It means errors left)
.IP \fB\-d\fP
Delete the specified file. If more that one file with that name exists, the
first one is deleted.
.IP \fB\-f\fP
Salvage unused cluster chains to files. By default, unused clusters are
added to the free disk space except in auto mode (\fB-a\fP).
.IP \fB\-l\fP
List path names of files being processed.
.IP \fB\-n\fP
No-operation mode: non-interactively check for errors, but don't write
anything to the filesystem.
.IP \fB\-r\fP
Interactively repair the file system. The user is asked for advice whenever
there is more than one approach to fix an inconsistency.
.IP \fB\-t\fP
Mark unreadable clusters as bad.
.IP \fB-u\fP
Try to undelete the specified file. \fBdosfsck\fP tries to allocate a chain
of contiguous unallocated clusters beginning with the start cluster of the
undeleted file.
.IP \fB\-v\fP
Verbose mode. Generates slightly more output.
.IP \fB\-V\fP
Perform a verification pass. The file system check is repeated after the
first run. The second pass should never report any fixable errors. It may
take considerably longer than the first pass, because the first pass may
have generated long list of modifications that have to be scanned for each
disk read.
.IP \fB\-w\fP
Write changes to disk immediately.
.IP \fB\-y\fP
Same as \fB\-a\fP (automatically repair filesystem) for compatibility
with other fsck tools.
.LP
If \fB\-a\fP and \fB\-r\fP are absent, the file system is only checked,
but not repaired.
.SH "EXIT STATUS"
.IP 0
.PD 0
No recoverable errors have been detected.
.IP 1
Recoverable errors have been detected or \fBdosfsck\fP has discovered an
internal inconsistency. And errors are corrected.
.IP 4
Errors are detected, but some of them are not corrected.
.IP 8
\fBdosfsck\fP exit during operation with error, or unmanageable situation.
.IP 16
Usage error. check usage using '\fBdosfsck --help'\fP.
.IP 32
File system checking canceled by user request.
.IP 64
It's not FAT filesystem. \fBdosfsck\fP did not access the file system.
.IP 128
System call like open/read/write/lseek error.
.SH FILES
.IP "FOUND.000, FOUND.001, ..."
.PD
When recovering orphan cluster, recovered file are stored in these directories.
Each directory of FOUND.NNN can include 10 thousand files which named 'FSCKNNNN.REC'.
.IP "FSCK0000.REC, FSCK0001.REC, ..."
When recovering from a corrupted file system, \fBdosfsck\fP dumps recovered data
into files named 'FSCKNNNN.REC' in FOUND.NNN which is in the top level
directory of the file system.
.IP "FSCK0000.000, FSCK0000.001, ..."
When \fBdosfsck\fP found bad name entry, will rename those files to 'FSCKNNNN.NNN'
in entry's parent directory.
.SH BUGS
Should give more diagnostic messages. Undeleting files should use a more
sophisticated algorithm.
.SH "SEE ALSO"
.BR mkdosfs(8)
.BR dosfslabel(8)
.BR dosfsdump(8)
.SH AUTHORS
\fBdosfstools\fP were written by Werner Almesberger <werner.almesberger@lrc.di.epfl.ch>
Extensions (FAT32, VFAT) by Roman Hodek <roman@hodek.net>
fatprogs maintainer is JaeHoon Sim (jay.sim@lge.com).
.SH ACKNOWLEDGMENTS
.B fatprogs
is based on code from dofstools v2.11(GPLv2).


CHANGES
fatprogs v2.13.0 - released 2022-7-18
=====================================

CHANGES:
 * dosfsck: dirty flag clean after all changes are flushed.
 * dosfsck: remove 'reclaim_bitmap' field while reuse 'bitmap' for reclaiming.
 * dosfsck: change default action(removal) for corrupted LFN.
 * dosfsck: modify file size field instead of trucating file.
 * dosfsdump: make default option to dump metadata only.
 * dosfsck: consider corrupted directory if it has too many bad files.

NEW FEATURES:
 * dosfsdump: add option to redirect dump data to standard output.
 * dosfsck: use FOUND.XXX directory for reclaimed files.

BUF FIXES:
 * mkdosfs: fix wrong root directory initialization.
 * dosfsck: check boot signature more strictly to distinguish from ntfs.
 * dosfsck: remove checking time field of dot/dotdot entry.
 * mkdosfs: fix wrong media type location writing.
 * dosfsck: fix FAT start offset for unaligned (with PAGE_SIZE) FAT volume.
 * dosfsck: set EOC to first free FAT entry of root directory.


fatprogs v2.12.2 - released 2022-5-08
=====================================

CHANGES:
 * change version string definition.
 * several bugs fixed.


fatprogs v2.12.1 - released 2022-4-30
=====================================

CHANGES:
 * dosfsck: minimize memory using bitmap.
 * dosfsck: improve performance on changed list, FAT read, etc.
 * mkdosfs: minimize memory for FAT read.
 * Makefile: add target for debugging with address sanitizer.
 * Apply pread/pwrite instead of lseek and read/write.

NEW FEATURES:
 * dosfsdump: add dump tool for debugging.
 * dosfsck: read FAT from mmap memory.

BUG FIXES:
 * dosfsck: fix wrong calculated FAT size to read.

Move version.h to 'include' directory.


fatprogs v2.12.0 - released 2022-3-30
=====================================
Since fatprogs starts from dosfstools(v2.11) project, decide that
the version also starts with the last version.

CHANGES:
 * Apply last GPL v2 licensed version of dosfstools (v2.11)
 * Apply patches of distributor released under GPL v2 license.
 * Seperate source and header and put all sources in one directory.

NEW FEATURES:
 * dosfsck: Add handling volume label in boot sector and volume entry in
 root directory entry.
 * dosfsck: Add checking routine of dot/dotdot entries.
 * dosfsck: Add handling volume dirty flag.

BUG FIXES:
 * dosfsck: Fix memory leaks / stack smashing using address sanitizer.
 * dosfsck: Fix wrong truncation size of corrupted cluster.
 * dosfsck: Fix using unproper function.(sprintf -> memcpy)
 * dosfsck: Remove unnecessary LFN when it rename short filename.
 * dosfsck: Free lfn resource in memory in interactive mode.
 * dosfsck: Fix possible corruption due to multiple FAT.


src/dosfsck.c
/* SPDX-License-Identifier : GPL-2.0 */

/* dosfsck.c  -  User interface */

/* Written 1993 by Werner Almesberger */

/* FAT32, VFAT, Atari format support, and various fixes additions May 1998
 * by Roman Hodek <Roman.Hodek@informatik.uni-erlangen.de> */

#include "version.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>

#include "common.h"
#include "dosfsck.h"
#include "io.h"
#include "boot.h"
#include "fat.h"
#include "file.h"
#include "check.h"

int interactive = 0, list = 0, test = 0, verbose = 0, write_immed = 0;
int check_dirty_only = 0;
int atari_format = 0;
int remain_dirty = 0;
unsigned n_files = 0;
void *mem_queue = NULL;

uint32_t max_clus_num;

static void usage(char *name)
{
    fprintf(stderr, "usage: %s [-aAflrtvVwy] [-d path -d ...] "
            "[-u path -u ...]\n%15sdevice\n", name, "");
    fprintf(stderr, "  -a       automatically repair the file system\n");
    fprintf(stderr, "  -A       toggle Atari file system format\n");
    fprintf(stderr, "  -C       only check filesystem dirty flag(FAT32/16 only)\n");
    fprintf(stderr, "  -d path  drop that file\n");
    fprintf(stderr, "  -f       salvage unused chains to files\n");
    fprintf(stderr, "  -l       list path names\n");
    fprintf(stderr, "  -n       no-op, check non-interactively without changing\n");
    fprintf(stderr, "  -r       interactively repair the file system\n");
    fprintf(stderr, "  -t       test for bad clusters\n");
    fprintf(stderr, "  -u path  try to undelete that (non-directory) file\n");
    fprintf(stderr, "  -v       verbose mode\n");
    fprintf(stderr, "  -V       perform a verification pass\n");
    fprintf(stderr, "  -w       write changes to disk immediately\n");
    fprintf(stderr, "  -y       same as -a, for compat with other *fsck\n");
}

int main(int argc, char **argv)
{
    DOS_FS fs;
    int rw, salvage_files, verify, c;
    int ret = 0;
    int dirty_flag = 0;
    uint32_t free_clusters;

    salvage_files = verify = 0;
    rw = 1;
    interactive = 1;
    check_atari(&atari_format);

    while ((c = getopt(argc, argv, "AaCd:flnrtu:vVwy")) != EOF) {
        switch (c) {
            case 'A': /* toggle Atari format */
                atari_format = !atari_format;
                break;
            case 'a':
            case 'y':
                rw = 1;
                interactive = 0;
                salvage_files = 1;
                break;
            case 'C':
                check_dirty_only = 1;
                interactive = 0;
                break;
            case 'd':
                file_add(optarg, fdt_drop);
                break;
            case 'f':
                salvage_files = 1;
                break;
            case 'l':
                list = 1;
                break;
            case 'n':
                rw = 0;
                interactive = 0;
                break;
            case 'r':
                rw = 1;
                interactive = 1;
                break;
            case 't':
                test = 1;
                break;
            case 'u':
                file_add(optarg, fdt_undelete);
                break;
            case 'v':
                verbose = 1;
                break;
            case 'V':
                verify = 1;
                break;
            case 'w':
                write_immed = 1;
                break;
            default:
                usage(argv[0]);
                exit(EXIT_SYNTAX_ERROR);
        }
    }

    if ((test || write_immed) && !rw) {
        fprintf(stderr, "-t and -w require -a or -r\n");
        exit(EXIT_SYNTAX_ERROR);
    }

    if (optind != argc - 1) {
        usage(argv[0]);
        exit(EXIT_SYNTAX_ERROR);
    }

    printf("dosfsck " VERSION ", " VERSION_DATE ", FAT32, LFN\n");
    fs_open(argv[optind], rw);
    read_boot(&fs);

    if (verify)
        printf("\nStarting check/repair pass.\n");

    do {
        n_files = 0;
        dirty_flag = 0;
        read_fat(&fs);

        if ((fs.fat_bits == 32) || (fs.fat_bits == 16)) {
            dirty_flag = check_dirty_flag(&fs);
        }

        if (check_dirty_only) {
            if (dirty_flag) {
                if (verify)
                    printf("  Just check filesystem dirty flag, exit!\n");
                exit(EXIT_ERRORS_LEFT);
            }
            else {
                if (verify)
                    printf("  Filesystem dirty flag is clean. exit!\n");
                exit(EXIT_NO_ERRORS);
            }
        }

        ret = scan_root(&fs);
        if (ret) {
            qfree(&mem_queue);
        }
    } while (ret);

    if (test)
        fix_bad(&fs);

    check_volume_label(&fs);

    if (salvage_files)
        reclaim_file(&fs);
    else
        reclaim_free(&fs);

    free_clusters = update_free(&fs);
    file_unused();

    if (verbose) {
        print_mem();
#ifdef DEBUG
        print_changes();
#endif
    }
    qfree(&mem_queue);

    if (verify) {
        printf("\nStarting verification pass.\n");
        n_files = 0;
        read_fat(&fs);
        scan_root(&fs);
        check_volume_label(&fs);
        reclaim_free(&fs);
        if (verbose)
            print_mem();

        qfree(&mem_queue);
    }

    if (fs_changed()) {
        if (rw) {
            if (interactive)
                rw = get_key("yn", "Perform changes ? (y/n)") == 'y';
            else
                printf("\nPerforming changes.\n");
        }
        else
            printf("\nLeaving file system unchanged.\n");
    }

    printf("%s: %u files, %u/%u clusters\n", argv[optind], n_files,
            fs.clusters - free_clusters, fs.clusters);

    clean_boot(&fs);

    /* sync for modified data */
    ret = fs_flush(rw);

    if (!remain_dirty)
        clean_dirty_flag(&fs);

    if (fs.fat_cache.addr) {
        fs_munmap(fs.fat_cache.addr, FAT_CACHE_SIZE);
    }

    /* sync for dirty flag */
    fs_flush(rw);

    fs_close();
    if (remain_dirty)
        return EXIT_ERRORS_LEFT;

    return (ret ? EXIT_CORRECTED : EXIT_NO_ERRORS);
}

/* Local Variables: */
/* tab-width: 8     */
/* End:             */
